### DISCUSSION TITLE
Audio output jacks and switching to mono using GPIO pin

### DISCUSSION THREAD

**Discussion by dcoredump** [probonopd/MiniDexed#181] [Ideas]
Hi,

I am using my gear for my band and we are sometimes using mono monitor speakers for our rehearsal room and stereo speakers on stage.

When using MiniDexed with only one output connecting to the mono monitor I have to check if everything is panned to the "right" side. That's bad.

My idea: I want to connect two mono 6.3 jacks to my Hifiberry audio hat. Those jacks have an additinal NC contact to GND when nothing is connected. Both contacts could be connected to a GPIO pin (with pullup resistor) and inside MiniDexed at the end of the audio path a check could be implemented which activates a stereo-to-mono function and everything is fine.

I am doing something similar with MicroDexed and this is a big help for me.

What do YOU think about this?

**Comment by craigyjp** [comment]
Why not wait until multi performance modes are available then you can
create performance modes to suit your needs.

**Reply by dcoredump** [reply]
Yes, this is also an option. But we sometimes have not much time for getting our equipment up and running on stage (especially for me because I have to setup a synth and a guitar), so a simple stupid method for getting mono sounds when only one output jack is used is quite helpful.

I have started to make my MiniDexed hardware ready for this and will try it hopefully the next days. If it works it would be very nice if the code can go into the main branch.

For users who do not need a mono switch, it only uses one additional CPU cycle and for those who use it, the code for making mono from stereo very simple and with CMSIS5/NEON very fast and should not have any impact to the rest of the audio queue (adding (int16_t) right and left channel together (as int32_t) and shift it one bit right - thats all).

**Comment by probonopd** [comment]
I don't know how hard it would be to write a stereo2mono function. But wiring that function to a "button" wired to a GPIO should be trivial. Or have a Mono/Stereo menu item.

**Reply by dcoredump** [reply]
It's really trivial to make mono from stereo (the other way around us not so trivial 😀).

**Reply by dcoredump** [reply]
> I don't know how hard it would be to write a stereo2mono function. But wiring that function to a "button" wired to a GPIO should be trivial. Or have a Mono/Stereo menu item.

Hm, I havn't found an example yet. Do you have an idea how to manage that a rising edge on GPIO18 set a flag and a falling egde clears it?

Regards, Holger

**Reply by probonopd** [reply]
Ah, I thought there was a GPIO reading example in Circle but apparently there isn't yet. So I have [asked](https://github.com/rsta2/circle/issues/292) for one.

**Reply by probonopd** [reply]
> Generating interrupts from GPIO lines is shown in [sample/09-softserial](https://github.com/rsta2/circle/blob/master/sample/09-softserial/softserial.cpp). `m_RxDPin.Read()` returns the current input level (`LOW` (0) or `HIGH` (1)).

**Comment by craigyjp** [comment]
Would it not be easier to have multiple performance patches like the DX7s and MK2.

**Reply by probonopd** [reply]
I guess sooner or later someone should implement that, yes.

**Comment by CocoFBest** [comment]
It's a bit late, but you can do it via hardware.
Use 2 audio jacks with switching. Wire the right side so if no plug is attached the signal passes to the left/mono channel.
Thats how it's done in practically every audio application.
https://www.cuidevices.com/blog/understanding-audio-jack-switches-and-schematics
Regards
